sklearn/preprocessing/discretization.py

```diff
@@ -199,15 +221,31 @@ def transform(self, X):
 
         Returns
         -------
-        Xt : numeric array-like, shape (n_samples, n_features)
+        Xt : numeric array-like
             Data in the binned space.
         """
         check_is_fitted(self, ["offset_", "bin_width_"])
         X = self._validate_X_post_fit(X)
 
-        return _transform_selected(X, self._transform,
-                                   self.transformed_features_, copy=True,
-                                   retain_order=True)
+        Xt = _transform_selected(X, self._transform,
+                                 self.transformed_features_, copy=True,
+                                 retain_order=True)
+
+        # only one-hot encode discretized features
+        mask = np.ones(X.shape[1], dtype=bool)
+        if self.ignored_features is not None:
+            mask[self.ignored_features] = False
+
+        if self.encode == 'onehot':
+            return OneHotEncoder(n_values=np.array(self.n_bins_)[mask],
+                                 categorical_features=mask,
```

Comment:
maybe should use logic like `categorical_features=None if self.ignored_features is None else mask` just to not waste time in OHE